File: backend/app/api/notifications.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime

from app.auth.jwt import get_current_user
from app.services.notification_service import notification_service
from app.services.alert_engine import alert_engine, AlertRule, AlertType, AlertCondition

router = APIRouter()


class TelegramConfig(BaseModel):
    bot_token: str
    chat_id: str


class WebhookConfig(BaseModel):
    name: str
    url: str
    events: List[str]


class AlertRuleCreate(BaseModel):
    alert_type: AlertType
    symbol: Optional[str]
    condition: AlertCondition
    value: float
    channel: str = "telegram"
    cooldown_seconds: int = 300


class AlertRuleResponse(BaseModel):
    id: str
    alert_type: str
    symbol: Optional[str]
    condition: str
    value: float
    channel: str
    is_active: bool
    last_triggered: Optional[datetime]


@router.post("/telegram/configure")
async def configure_telegram(config: TelegramConfig, user=Depends(get_current_user)):
    notification_service.configure_telegram(config.bot_token, config.chat_id)
    return {"status": "configured", "message": "Telegram bot configured successfully"}


@router.post("/telegram/test")
async def test_telegram(user=Depends(get_current_user)):
    success = await notification_service.send_telegram(
        "🧪 <b>MT5 Router Test</b>\nTelegram notifications are working!"
    )
    return {"success": success}


@router.post("/webhooks")
async def add_webhook(config: WebhookConfig, user=Depends(get_current_user)):
    notification_service.add_webhook(config.name, config.url, config.events)
    return {"status": "added", "name": config.name}


@router.get("/webhooks")
async def list_webhooks(user=Depends(get_current_user)):
    return notification_service.webhook_urls


@router.delete("/webhooks/{name}")
async def delete_webhook(name: str, user=Depends(get_current_user)):
    if name in notification_service.webhook_urls:
        del notification_service.webhook_urls[name]
        return {"status": "deleted"}
    raise HTTPException(status_code=404, detail="Webhook not found")


@router.post("/alerts")
async def create_alert(rule: AlertRuleCreate, user=Depends(get_current_user)):
    import uuid

    alert_id = str(uuid.uuid4())[:8]

    alert_rule = AlertRule(
        id=alert_id,
        user_id=user.id,
        alert_type=rule.alert_type,
        symbol=rule.symbol,
        condition=rule.condition,
        value=rule.value,
        channel=rule.channel,
        is_active=True,
        cooldown_seconds=rule.cooldown_seconds,
    )

    alert_engine.add_rule(alert_rule)

    return {
        "id": alert_id,
        "status": "created",
        "rule": {
            "type": rule.alert_type,
            "symbol": rule.symbol,
            "condition": f"{rule.condition} {rule.value}",
            "channel": rule.channel,
        },
    }


@router.get("/alerts")
async def list_alerts(user=Depends(get_current_user)):
    user_rules = [
        {
            "id": rule_id,
            "type": rule.alert_type.value,
            "symbol": rule.symbol,
            "condition": rule.condition.value,
            "value": rule.value,
            "channel": rule.channel,
            "is_active": rule.is_active,
            "last_triggered": rule.last_triggered,
        }
        for rule_id, rule in alert_engine.rules.items()
        if rule.user_id == user.id
    ]
    return user_rules


@router.put("/alerts/{alert_id}")
async def update_alert(
    alert_id: str, is_active: Optional[bool] = None, user=Depends(get_current_user)
):
    if alert_id not in alert_engine.rules:
        raise HTTPException(status_code=404, detail="Alert not found")

    if is_active is not None:
        alert_engine.update_rule(alert_id, is_active=is_active)

    return {"status": "updated"}


@router.delete("/alerts/{alert_id}")
async def delete_alert(alert_id: str, user=Depends(get_current_user)):
    alert_engine.remove_rule(alert_id)
    return {"status": "deleted"}


@router.post("/test")
async def send_test_notification(
    channel: str = "telegram", user=Depends(get_current_user)
):
    results = await notification_service.notify(
        event="test",
        data={"message": "Test notification from MT5 Router"},
        channels=[channel],
    )
    return {"results": results}

File: frontend/src/api/notifications.ts
import { api } from './client'

export interface TelegramConfig {
  bot_token: string
  chat_id: string
}

export interface WebhookConfig {
  name: string
  url: string
  events: string[]
}

export interface AlertRule {
  id: string
  type: string
  symbol: string | null
  condition: string
  value: number
  channel: string
  is_active: boolean
  last_triggered: string | null
}

export interface AlertRuleCreate {
  alert_type: string
  symbol?: string
  condition: string
  value: number
  channel?: string
  cooldown_seconds?: number
}

export const notificationsApi = {
  // Telegram
  configureTelegram: (config: TelegramConfig) =>
    api.post('/notifications/telegram/configure', config),
  
  testTelegram: () =>
    api.post('/notifications/telegram/test'),
  
  // Webhooks
  addWebhook: (config: WebhookConfig) =>
    api.post('/notifications/webhooks', config),
  
  listWebhooks: () =>
    api.get<Record<string, any>>('/notifications/webhooks'),
  
  deleteWebhook: (name: string) =>
    api.delete(`/notifications/webhooks/${name}`),
  
  // Alerts
  createAlert: (alert: AlertRuleCreate) =>
    api.post<AlertRule>('/notifications/alerts', alert),
  
  listAlerts: () =>
    api.get<AlertRule[]>('/notifications/alerts'),
  
  updateAlert: (id: string, is_active: boolean) =>
    api.put(`/notifications/alerts/${id}`, { is_active }),
  
  deleteAlert: (id: string) =>
    api.delete(`/notifications/alerts/${id}`),
  
  // Test
  sendTestNotification: (channel: string = 'telegram') =>
    api.post(`/notifications/test?channel=${channel}`),
}
